go/store/nbs/file_manifest.go

```diff
@@ -33,21 +34,119 @@ import (
 
 	"github.com/dolthub/fslock"
 
+	"github.com/dolthub/dolt/go/store/chunks"
 	"github.com/dolthub/dolt/go/store/hash"
 	"github.com/dolthub/dolt/go/store/util/tempfiles"
 )
 
 const (
 	manifestFileName = "manifest"
 	lockFileName     = "LOCK"
+
+	storageVersion4 = "4"
+
+	prefixLen = 5
 )
 
-// fileManifest provides access to a NomsBlockStore manifest stored on disk in |dir|. The format
-// is currently human readable:
+type manifestParser func(r io.Reader) (manifestContents, error)
+type manifestWriter func(temp io.Writer, contents manifestContents) error
+type manifestChecker func(upstream, contents manifestContents) error
+
+// ParseManifest parses s a manifest file from the supplied reader
+func ParseManifest(r io.Reader) (ManifestInfo, error) {
+	fm5 := fileManifestv5{}
+	return fm5.parseManifest(r)
+}
+
+func MaybeMigrateFileManifest(ctx context.Context, dir string) (bool, error) {
+	f, err := openIfExists(filepath.Join(dir, manifestFileName))
+	if err != nil {
+		return false, err
+	}
+	if f == nil {
+		// no manifest exists, no need to migrate
+		return false, nil
+	}
+	defer func() {
+		err = f.Close()
+	}()
+
+	fm5 := fileManifestv5{dir}
+	ok, _, err := fm5.ParseIfExists(ctx, &Stats{}, nil)
+	if ok && err == nil {
+		// on v5, no need to migrate
+		return false, nil
+	}
+
+	fm4 := fileManifestV4{dir}
+	ok, contents, err := fm4.ParseIfExists(ctx, &Stats{}, nil)
+	if err != nil {
+		return false, err
+	}
+	if !ok {
+		// expected v4 or v5
+		return false, fmt.Errorf("could not read file manifest")
+	}
+
+	check := func(upstream, contents manifestContents) error {
+		if upstream.gcGen == contents.gcGen {
+			return errors.New("error migrating manifest")
+		}
+		return nil
+	}
+	contents.gcGen = contents.lock
+
+	_, err = updateWithParseWriterAndChecker(ctx, dir, fm5.writeManifest, fm4.parseManifest, check, contents.lock, contents, nil)
+
+	if err != nil {
+		return false, nil
+	}
+
+	// clear manifest cache
+	makeGlobalCaches()
```

Comment:
These seems likely to violate concurrency expectations around global caches? In particular, was this behind a `sync.Once` before?

Reply:
Yes, all other instances were behind a sync.Once. I'll remove this